### PR TITLE
NO-TICKET: Fix session

### PR DESCRIPTION
**Pull request by TomasChladekSL**
### Description

- Remove `sessionManager.reset()` from `SplunkRumAgentCore.install()` so a valid persisted session is reused instead of creating a new one on every process start.
- Add unit test verifying `install()` reuses an existing valid session without emitting `onSessionChanged`.

### Checklist

- [x] My code follows the project's coding standards.
- [x] I have run linters/formatters and fixed any issues.
- [x] There are no merge conflicts.
- [x] I have performed a self-review of my code.
- [x] All new and existing tests pass locally.
- [x] I have added license headers to all files.
- [x] (If applicable) I have added unit tests for my changes.
- [ ] (If applicable) I have updated the sample app for integration testing.
- [ ] (If applicable) I have updated any relevant documentation.

### Generative AI usage

- [ ] GAI was not used (or, no additional notation is required)
- [ ] Coder created a draft manually that was non-substantively modified by GAI (e.g., refactoring was performed by GAI on manually written code)
- [x] GAI was used to create a draft that was subsequently customized or modified
- [ ] Code was generated entirely by GAI

### How to Test These Changes

Run "install reuses existing valid session" test.

### CODE REVIEW

**Review by tonzhan2 — APPROVED** [review]

File: integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt [hunk 1]
```diff
@@ -79,8 +79,6 @@ internal object SplunkRumAgentCore {
 
         Logger.d(TAG, "install(agentConfiguration: $agentConfiguration, moduleConfigurations: $moduleConfigurations)")
 
-        sessionManager.reset()
-
         val storage = AgentStorage.attach(application)
```
[the remark this answers is not in the document]
Comment (surbhiia) [reply integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt:82]: @TomasChladekSL I've unresolved this as this is not fixed yet in the PR. Do we have a ticket to resolve it but also I think it should be resolved in this PR to keep sampling decision consistent across multiple process re-starts which don't change session? 

```
 // Sampling.
        val shouldBeRunning = when (val samplingRate = agentConfiguration.session.samplingRate.coerceIn(0.0, 1.0)) {
            0.0 -> false
            1.0 -> true
            else -> Math.random() < samplingRate
        }

        if (!shouldBeRunning) return OpenTelemetry.noop()
```

File: integration/sessionreplay/src/main/kotlin/com/splunk/rum/integration/sessionreplay/SessionReplayModuleIntegration.kt [hunk 2]
```diff
@@ -80,10 +80,8 @@ internal object SessionReplayModuleIntegration : ModuleIntegration<SessionReplay
 
         isInstalled = true
 
-        if (isPendingSessionChange) {
-            isPendingSessionChange = false
-            processSessionChange()
-        }
+        isPendingSessionChange = false
```
Comment (surbhiia) [inline integration/sessionreplay/src/main/kotlin/com/splunk/rum/integration/sessionreplay/SessionReplayModuleIntegration.kt]: `isPendingSessionChange` is not used anywhere else, it was only used here, so now we can remove this variable if not needed anymore.

Comment (surbhiia) [reply integration/sessionreplay/src/main/kotlin/com/splunk/rum/integration/sessionreplay/SessionReplayModuleIntegration.kt]: Also note that - if `onSessionChange()` below is called before SessionReplayModule installation, `processSessionChange()` will be called twice, is that acceptable?

Comment (TomasChladekSL) [reply integration/sessionreplay/src/main/kotlin/com/splunk/rum/integration/sessionreplay/SessionReplayModuleIntegration.kt]: > `isPendingSessionChange` is not used anywhere else, it was only used here, so now we can remove this variable if not needed anymore.

Removed unused property. 

> Also note that - if `onSessionChange()` below is called before SessionReplayModule installation, `processSessionChange()` will be called twice, is that acceptable?

Even if `onSessionChange()` is called before the module is installed, `processSessionChange()` won't actually run twice. On that early call, `isInstalled` is still `false`, so it hits the `if (!isInstalled) return` guard and short-circuits immediately as a no-op.

File: integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (TomasChladekSL) [reply integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt:82]: This PR was intended only as a quick fix for an issue that is currently affecting one of our customers. Properly addressing the sampling consistency across process restarts would require more extensive testing, so we'll handle it in a follow-up rather than in this PR.

**Review by surbhiia — APPROVED** [review]

File: integration/agent/internal/src/main/kotlin/com/splunk/rum/integration/agent/internal/sampling/SessionSampler.kt [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+
+    /**
+     * Returns whether the session identified by [sessionId] should be sampled in for the given
+     * [samplingRate] (clamped to `0.0..1.0`).
+     */
+    fun shouldSample(samplingRate: Double, domain: String = "", sessionId: () -> String): Boolean =
+        when (val rate = samplingRate.coerceIn(0.0, 1.0)) {
+            0.0 -> false
+            1.0 -> true
+            else -> (sessionIdToUInt32(sessionId()) xor domainSalt(domain)) < (rate * UINT32_MAX).toLong()
```
[the remark this answers is not in the document]
Comment (surbhiia) [reply integration/agent/internal/src/main/kotlin/com/splunk/rum/integration/agent/internal/sampling/SessionSampler.kt]: @TomasChladekSL We should perhaps correct this:

The domain salt approach still correlates agent and Session Replay sampling because XOR only rearranges the original 32-bit value; it does not re-randomize it. After conditioning on sessions that passed agent sampling, the Replay distribution may be heavily biased—for example, agent rate 0.5 with Replay rate 0.2 can select no Replay sessions.
I suggest deriving an independent deterministic value for each sampling domain, for example:
SHA-256(domain + "\u0000" + sessionId)
Use the first four digest bytes as an unsigned 32-bit value and compare them with the sampling threshold. Domains such as "agent" and "sessionReplay" would then produce stable but effectively independent decisions. This requires no additional dependency, remains stable across process restarts, and also handles non-hex session IDs correctly.
We should add a joint-distribution test confirming that, for agent rate 0.5 and Replay rate 0.2, approximately 20% of agent-sampled sessions are selected for Replay—not 0%, 40%, or 100%.

File: integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (surbhiia) [reply integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt:82]: @TomasChladekSL This still appears possible on the latest commit. When reusing a valid persisted session, could we ensure that SESSION_IDS contains an entry for the current ID? If missing, backfill it using sessionValidUntil - maxSessionLength as validFrom and persist the repaired history. We should add a test with a valid SESSION_ID/valid-until but empty history and verify that sessionId(now) returns the reused ID without throwing. Since this is an SDK runtime telemetry path, it would also be preferable for sessionId(timestamp) to degrade safely rather than throw for recoverable storage inconsistencies.

File: integration/agent/internal/src/main/kotlin/com/splunk/rum/integration/agent/internal/sampling/SessionSampler.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (TomasChladekSL) [reply integration/agent/internal/src/main/kotlin/com/splunk/rum/integration/agent/internal/sampling/SessionSampler.kt]: Fixed

File: integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt [hunk 4]
```diff
@@ -64,11 +65,10 @@ internal object SplunkRumAgentCore {
         globalAttributes: MutableAttributes,
         offlineOtelDataProcessor: OfflineOtelDataProcessor
     ): OpenTelemetry {
-        // Sampling.
-        val shouldBeRunning = when (val samplingRate = agentConfiguration.session.samplingRate.coerceIn(0.0, 1.0)) {
-            0.0 -> false
-            1.0 -> true
-            else -> Math.random() < samplingRate
+        sessionManager.attachLifecycleObserver(application)
```
[the remark this answers is not in the document]
Comment (TomasChladekSL) [reply integration/agent/api/src/main/kotlin/com/splunk/rum/integration/agent/api/internal/SplunkRumAgentCore.kt:68]: `install()` have to be called only from main thread.

**Comment by TomasChladekSL** [comment]
All comments are addressed. Session management is a candidate for a rewrite on the next pass, it’s been edited by three people with different requirements and has become spaghetti with side effects.
